text/0067-sim-initial.md

```diff
@@ -0,0 +1,120 @@
+        ctx.set(o, a_val * b_val)
+```
+
+It creates a trigger object that fires on every change of `a` and `b`, ensuring that `o = a * b`. However, if the initial state of `a`, `b`, and `o` is not consistent (eg. the first line is changed to `a = Signal(16, init=2)`), the trigger will not fire, and the process will fail to perform its function.
+
+It is not easily possible to fix this problem — `ctx.get()` cannot be used in a process and the only way to obtain signal values involves waiting on a trigger object. Further, even if `ctx.get()` was allowed in processes, solving the problem in the obvious way requires code duplication:
+
+```
+async def process(ctx):
+    ctx.set(o, ctx.get(a) * ctx.get(b))
+    async for a_val, b_val in ctx.changed(a, b):
+        ctx.set(o, a_val * b_val)
+```
+
+A similar problem also appears for flip-flops when asynchronous resets are involved. Consider the following code:
+
+```
+d = Signal()
+q = Signal()
+
+async def ff_process(ctx):
+    async for clk_edge, rst_val, d_val in ctx.tick().sample(d):
+        if rst_val:
+            ctx.set(q, 1)
+        elif clk_hit:
+            ctx.set(q, d_val)
+```
+
+While the code is meant to implement an asynchronous-set flip-flop, it fails to do so correctly when the `sync` domain has an asynchronous reset and the reset is initially asserted.
+
+
+## Guide-level and reference-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When a trigger object is used in an `async for` loop, the `.initial()` method can be used to ensure the loop body executes once immediately, in addition to any other triggers:
+
+```
+async def process(ctx):
+    async for a_val, b_val in ctx.changed(a, b).initial():
+        # Executed immediately after starting the process, then executed again every time `a` or `b` changes.
+        ctx.set(o, a_val * b_val)
+```
+
+```
+async def ff_process(ctx):
+    async for clk_edge, rst_val, d_val in ctx.tick().sample(d).initial():
+        # Executed immediately after starting the process if reset is asynchronous and initially asserted;
+        # then executed on every clock rising edge and (if reset is asynchronous) every reset rising edge
+        if rst_val:
+            ctx.set(q, 1)
+        elif clk_hit:
+            ctx.set(q, d_val)
+```
+
+The `.initial()` method is meant only for use with `async for` — passing the result of `.initial()` to plain `await` is an error.
+
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Both combinable and initial trigger objects are amended with one new method:
+
+- `initial()` 
+  - Create a new trigger object of the same type by copying the current object and appending an initial trigger. This causes the resulting trigger object to fire once immediately after first starting iteration by an `async for` loop. Awaiting such a trigger with plain `await` is an error. The `.initial()` trigger doesn't contribute any components to the yielded value.
```

Comment:
It is obvious what `.changed(...).initial()` would return. What about `.tick(...).initial()` though? Can it race with other processes at time 0?

Reply:
I don't know what kind of races you mean; all races should be prevented by other processes' changes not happening until the next delta cycle, right?

Reply:
It's a little more complicated than that. Normally, races (e.g. pulse on clk and rst simultaneously) are avoided because async signals don't toggle simultaneously with other signals. For example, clocks will be usually toggled by `add_clock` while everything else in the design is waiting for that tick, and resets will be usually asserted by testbenches as `ctx.set(rst, 1)` that doesn't change any other signal unless you opt into it via `ctx.set(Cat(...), ...)`.

However, things are weirder at time 0. Consider a situation where many combinational processes are used at time 0, all with `.initial()` or equivalent. These will all run in the same delta cycle, and may change a large amount of signals in that cycle. If several of those signals are a part of the same async control set you get a race.

In short, we don't actually *prevent* races, we make them harder to accidentally trigger via UI design, and in case of `.initial()` specifically the UI design part isn't there.

---

So knowing exactly what `.tick()` returns is important because the behavior during a race should be well-defined. Is it neither `clk` nor `rst? Is it `rst` but not `clk`? On careful reading I conclude that it's the latter but the RFC text should definitely say that explicitly.